[xumarcus] iP

xumarcus opened the pull request:
# Zoe
Features:
- text-based
- easy to learn

Supported tasks:
1. Todo
2. Deadline
3. Event

```java
try (Reader reader = new FileReader(path)) {
    List<Task> tasks = gson.fromJson(reader, new TypeToken<List<Task>>() {
    }.getType());
    if (tasks == null) throw new ZoeIOException("File is empty");
    return tasks;
}
```
Coming soon:
- [ ] Reminders

:grinning:
> Trivia: In leetspeak, 2103 corresponds to z1oe.

Zoe uses [gson](https://github.com/google/gson), like `gson.toJson(taskList, writer)`, to **store** and *load* tasks, ~~i.e.~~.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Level-0

Commit:
Level-1

Commit:
Level-2

Commit:
Level-3

Commit:
Level-4

Commit:
A-TextUiTesting

Commit:
Level-5

Commit:
Level-5

Commit:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	src/main/java/TodoExecutor.java
#	src/main/java/Zoe.java
#	text-ui-test/EXPECTED.TXT
#	text-ui-test/input.txt

Commit:
Level-6

Commit:
A-Enums

Commit:
Add lombok dependency

Commit:
Add Gson support

Commit:
Add read/save and resource API for tasks

Commit:
Use resource API for mark and unmarking tasks

Commit:
Refactor remove to use index

Commit:
Update test script to include external dependencies

Commit:
Fix tests

Commit:
Fix lombok warning

Commit:
Add dependencies to lib directory to satisfy iP requirements

Commit:
Remove taskName from Task (but not its inheritors)

Commit:
Understand dates in command

Commit:
Update tests

Commit:
Merge branch 'branch-Level-8'

Commit:
Refactor to Command pattern

Commit:
Extract storage class

Commit:
Rename ToDo to Todo

Commit:
Add LocalDate support for Gson

Commit:
Extract taskList class

Commit:
Extract parser class

Commit:
Extract ui class

Commit:
Refactor the code to extract out closely related code as classes

Commit:
Update tests

Commit:
Merge branch 'branch-A-MoreOOP'

Commit:
Divide into multiple packages

Commit:
Merge branch 'branch-A-Packages'

Commit:
Merge branch 'add-gradle-support'

Commit:
Set up gradle

Commit:
Fix spacing bug and access specifiers for Parser

Commit:
Add getters to commands for testing

Commit:
Add test for Parser

Commit:
Move to src/test/java

Commit:
Add test for task formatting

Commit:
Merge branch 'branch-A-JUnit'

Commit:
A-Jar

Commit:
Add JavaDoc comments

Commit:
Tweak the code to comply with the given coding standard

Commit:
Remove lombok for derived classes

Commit:
Add ui find/list message

Commit:
Add FindCommand

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/zoe/command/Command.java
#	src/main/java/zoe/task/Task.java

Commit:
Rearrange code

Commit:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/zoe/task/Event.java
#	src/main/java/zoe/task/Task.java
#	src/main/java/zoe/task/Todo.java
#	src/main/java/zoe/ui/Ui.java

Commit:
Fix merge conflicts

Commit:
Rearrange code

Commit:
Add checkstyle

Commit:
Merge branch 'branch-A-CheckStyle'

Commit:
Implement JavaFX tutorial

Commit:
Integrate tutorial GUI into chatbot

Commit:
Merge branch 'branch-Level-10'

varuuuun commented on src/main/java/zoe/adapters/gson/LocalDateTypeAdapter.java:
```diff
@@ -0,0 +1,27 @@
+
+// https://mkyong.com/java/gson-supports-java-8-date-time-types/#default-gson-does-not-support-java-8-date-time-type
+public class LocalDateTypeAdapter implements JsonSerializer<LocalDate>, JsonDeserializer<LocalDate> {
+    private static final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
```
`formatter` isnt in capital case, should it be in SCREAMING_SNAKE_CASE?

xumarcus replied:
Need to clarify. Most code samples use lower case

varuuuun commented on src/main/java/zoe/ZoeException.java:
```diff
@@ -0,0 +1,7 @@
+package zoe;
+
+public class ZoeException extends Exception {
```
What is this error for? Some javadoc here would be nice. Especially since you have 2 error classes, to differentiate the use cases.

varuuuun commented on src/main/java/zoe/ZoeIOException.java:
```diff
@@ -0,0 +1,7 @@
+package zoe;
+
+public class ZoeIOException extends ZoeException {
```
What is this error for? Some javadoc here would be nice. Especially since you have 2 error classes, to differentiate the use cases.

Commit:
Bump version from "v0.1" to "v0.2"

The shadowJar archive version was previously set to "v0.1", which no longer accurately reflects the current state of the build artifacts. This version bump is necessary to mark the incremental changes and improvements made since the last release.

Let's update the archiveVersion to "v0.2" to ensure clear versioning and better traceability of releases. This change will help maintain consistency across the build system and prevent potential confusion during deployment.

Further version bumps and updates will follow as additional changes are integrated into the project.

Commit:
Refactor code to resolve Checkstyle violations

The codebase contains multiple Checkstyle violations, including line
length, missing Javadoc comments, and inconsistent
indentation. These violations cause the build to fail when Checkstyle
checks are enabled and reduce code readability.

Enforcing Checkstyle rules ensures consistent coding standards, making
the code easier to maintain and reducing noise in future code reviews.

Commit:
Merge branch 'branch-A-FullCommitMessage'

Commit:
Use the Streams feature of Java

Commit:
Use assert feature to document important assumptions

Commit:
Examines the code and refactor to improve the code quality where necessary critically

xumarcus committed:
Merge pull request #1 from xumarcus/branch-A-Streams

Use the Streams feature of Java

Commit:
Merge remote-tracking branch 'origin/master' into branch-A-Assertions

xumarcus committed:
Merge pull request #3 from xumarcus/branch-A-Assertions

Use assert feature to document important assumptions

Commit:
Merge branch 'master' into branch-A-CodeQuality

xumarcus committed:
Merge pull request #2 from xumarcus/branch-A-CodeQuality

Examines the code and refactor to improve the code quality where nece…

Commit:
Make the command syntax more flexible

Commit:
Merge remote-tracking branch 'origin/branch-C-FriendlierSyntax'

Commit:
Remove unused import

varuuuun commented on src/main/java/zoe/Zoe.java:
```diff
@@ -0,0 +1,81 @@
+import zoe.command.Parser;
+import zoe.storage.Storage;
+import zoe.task.TaskList;
+import zoe.ui.Ui;
```
The order of your imports does not follow the required standards:
![image](https://github.com/user-attachments/assets/4a51d3fd-46d1-4c15-84e4-8a7f6fc5d06b)

xumarcus replied:
Fixed

varuuuun commented on src/main/java/zoe/storage/Storage.java:
```diff
@@ -0,0 +1,58 @@
+@AllArgsConstructor
+public class Storage {
+    // Call registerTypeAdapter(LocalDate.class, new zoe.adapters.gson.LocalDateTypeAdapter()) in zoe.adapters.gson.TaskTypeAdapter!
+    private final static Gson gson = new GsonBuilder()
```
Should this variable name be in all caps, to align with the SCREAMING_SNAKE_CASE requirement?

xumarcus replied:
This has side effect, so probably not

varuuuun reviewed (commented):
Really interesting work, some minor nits found.  

Generally your import statements across all files do not seem to follow the coding standard, refer to the comment I left under Zoe.java file. You might want to improve on that.

sreelakshmiharidos commented on src/main/java/zoe/storage/Storage.java:
```diff
@@ -0,0 +1,93 @@
+            }.getType());
+            if (tasks == null) {
+                throw new ZoeIoException("File is empty");
+            }
```
Perhaps you don't need to throw an exception for empty file, since it's a valid state. Instead, maybe just return an empty list and hence can reduce complexity in try-catch block.

sreelakshmiharidos commented on src/main/java/zoe/ui/Ui.java:
```diff
@@ -0,0 +1,165 @@
+    @FXML
+    private TextField userInput;
+    @FXML
+    @SuppressWarnings("unused")
```
Perhaps add a comment with @SuppressWarnings to convey rationale and improve readability?

sreelakshmiharidos commented on src/main/java/zoe/ui/Ui.java:
```diff
@@ -0,0 +1,165 @@
+     *
+     * @param tasks The list of tasks to be displayed.
+     */
+    public void showList(List<? extends Task> tasks) {
+        reply("Here are the matching tasks in your list:\n"
+                + IntStream
+                .range(0, tasks.size())
+                .mapToObj(i -> String.format("%d. %s", i + 1, tasks.get(i).getFullDescription()))
+                .collect(Collectors.joining("\n")));
+    }
```
Maybe use multiline string literal instead of concatenation to have consistent formatting with the rest of the code, thereby making the code easier to read and maintain?
```suggestion
    public void showList(List<? extends Task> tasks) {
    String taskListStr = IntStream
            .range(0, tasks.size())
            .mapToObj(i -> String.format("%d. %s", i + 1, tasks.get(i).getFullDescription()))
            .collect(Collectors.joining("\n"));

    reply(String.format("""
            Here are the matching tasks in your list:
            %s""", taskListStr));
}
```

sreelakshmiharidos commented on src/main/java/zoe/command/MarkCommand.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Override
+    public void execute(Storage storage, TaskList taskList, Ui ui) throws ZoeException {
+        assert(0 <= index && index < taskList.size());
```
Perhaps use explicit check instead of assert as assert is meant for debugging and not runtime validation?
```suggestion
        if (index < 0 || index >= taskList.size()) {
    throw new ZoeException("Invalid task index: " + index);
}
```

sreelakshmiharidos reviewed (commented):
LGTM. Just a few nits to improve code quality even more!

Commit:
Add a representative screenshot of the product to the docs folder

Commit:
Update README.md for user guide

Commit:
Merge branch 'branch-A-UserGuide'

Commit:
Add docs for compliance

Commit:
Ensure the product can handle common errors gracefully

Commit:
Bump version for release
